sts/client.go

```diff
@@ -0,0 +1,113 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sts
+
+import (
+	"context"
+	"crypto/tls"
+	"net/url"
+	"time"
+
+	"github.com/vmware/govmomi/sts/internal"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/soap"
+)
+
+// Client is a soap.Client targeting the STS (Secure Token Service) API endpoint.
+type Client struct {
+	*soap.Client
+}
+
+// DefaultPath is the default URL Path for the STS service
+var DefaultPath = "/sts/STSService"
+
+// NewClient returns an STS client targeting DefaultPath.
+func NewClient(ctx context.Context, c *vim25.Client) (*Client, error) {
+	sc := c.Client.NewServiceClient(DefaultPath, "urn:oasis:names:tc:SAML:2.0:assertion")
+
+	return &Client{sc}, nil
+}
+
+// TokenRequest parameters for issuing a SAML token.
+// At least one of Userinfo or Certificate must be specified.
+type TokenRequest struct {
+	Userinfo    *url.Userinfo    // Userinfo when set issues a Bearer token
+	Certificate *tls.Certificate // Certificate when set issues a HoK token
+	ActAs       string           // ActAs identity when set is delegated to the caller
+	Duration    time.Duration    // Duration is the token's lifetime, defaults to 1h
+	Renewable   bool             // Renewable allows the issued token to be renewed
+	Delegatable bool             // Delegatable allows the issued token to be delegated (e.g. for use with ActAs)
+}
+
+// Issue is used to request a security token.
+// The returned Signer can be used to sign SOAP requests, such as the SessionManager LoginByToken method and the RequestSecurityToken method itself.
+// See: http://docs.oasis-open.org/ws-sx/ws-trust/v1.4/errata01/os/ws-trust-1.4-errata01-os-complete.html#_Toc325658937
+func (c *Client) Issue(ctx context.Context, req TokenRequest) (*Signer, error) {
+	s := &Signer{
+		Certificate: req.Certificate,
+		user:        req.Userinfo,
+	}
+
+	if req.Duration == 0 {
+		req.Duration = time.Hour
+	}
+
+	created := time.Now().UTC()
+	rst := internal.RequestSecurityToken{
+		TokenType:          c.Namespace,
+		RequestType:        "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Issue",
+		SignatureAlgorithm: internal.SHA256,
+		Lifetime: &internal.Lifetime{
+			Created: internal.Time{created},
+			Expires: internal.Time{created.Add(req.Duration)},
+		},
+		Renewing: &internal.Renewing{
+			Allow: req.Renewable,
+			OK:    req.Renewable,
+		},
+		Delegatable: req.Delegatable,
+	}
+
+	if req.ActAs != "" {
+		rst.ActAs = &internal.ActAs{
+			Assertion: req.ActAs,
+		}
+	}
+
+	if s.Certificate == nil {
+		rst.KeyType = "http://docs.oasis-open.org/ws-sx/ws-trust/200512/Bearer"
+	} else {
```

Comment:
The documentation on the `TokenRequest` struct indicates that "At least one of Userinfo or Certificate must be specified."

This implementation suggests that a `Certificate`, if present, takes precedence. Should that be documented as well?

Relatedly, should we have a test that verifies the behavior when both `Userinfo` and `Certificate` are specified?